PubHub/PubHub.AdminPortal/Components/Auth/CustomAuthStateProvider.cs

```diff
@@ -1,20 +1,122 @@
-﻿using System.Security.Claims;
+﻿using System.Diagnostics;
+using System.IdentityModel.Tokens.Jwt;
+using System.Linq.Dynamic.Core.Tokenizer;
+using System.Security.Claims;
 using Blazored.LocalStorage;
-using Microsoft.AspNetCore.Authentication.JwtBearer;
 using Microsoft.AspNetCore.Components.Authorization;
+using Newtonsoft.Json.Linq;
+using PubHub.Common.Models.Authentication;
+using PubHub.Common.Services;
 
 namespace PubHub.AdminPortal.Components.Auth
 {
-    public class CustomAuthStateProvider 
+    public class CustomAuthStateProvider : AuthenticationStateProvider
     {
         private readonly ILocalStorageService _localStorage;
-        private readonly IHttpClientFactory _httpClientFactory;
-        private ClaimsPrincipal _anonymous = new ClaimsPrincipal(new ClaimsIdentity());
+        private readonly IAuthenticationService _authenticationService;
+        private readonly ClaimsPrincipal _anonymous = new(new ClaimsIdentity());
 
-        public CustomAuthStateProvider(ILocalStorageService localStorage, IHttpClientFactory httpClientFactory)
+        public CustomAuthStateProvider(ILocalStorageService localStorage, IAuthenticationService authenticationService)
         {
             _localStorage = localStorage;
-            _httpClientFactory = httpClientFactory;
+            _authenticationService = authenticationService;
+        }
+
+        public override async Task<AuthenticationState> GetAuthenticationStateAsync()
+        {
+            try
+            {
+                var token = await _localStorage.GetItemAsync<string>("token");
+                var refreshToken = await _localStorage.GetItemAsync<string>("refreshToken");
+
+                if (string.IsNullOrWhiteSpace(token) || string.IsNullOrWhiteSpace(refreshToken))
+                    return await Task.FromResult(new AuthenticationState(_anonymous));
+
+                TokenInfo tokenInfo = new() { Token = token, RefreshToken = refreshToken };                
+
+                var claimsPrincipal = GetClaims(token);
+
+                if (claimsPrincipal.FindFirst("exp") != null)
+                {
+                    var expireClaim = long.Parse(claimsPrincipal.FindFirst("exp")?.Value ?? "0");
+
+                    var utcExpire = DateTimeOffset.FromUnixTimeSeconds(expireClaim);
+                    var expiredate = utcExpire.DateTime;
+                    // If the token has expired, then we want to refresh the token and update the existing token and refresh token in our local storage.
+                    if (expiredate < DateTime.UtcNow)
+                    {
+                        var tokenResponse = await _authenticationService.RefreshTokenAsync(tokenInfo);
+                        if (tokenResponse != null && tokenResponse.Instance != null)
```

Comment:
If this check fails, will we then end up on line 65?

Reply:
I actually do not know what will happen in that case... I would think it just jumps down to line 58 and uses the already existing ones maybe. 🤔

Reply:
Reading the code I believe it would end up at line 65, however, I feel like that should already have been tested and confirmed. We want the behaviour we intend and not guess whether or not our code behaves as we'd like it to (:

Also; I assume the flow already has a `ClaimsPrincipal`, which means that we could potentially have a false positive in authentication on line 58, if that state wasn't reset earlier.
Otherwise it would hit 65.

But it's somewhat difficult to see when looking at code like this.